egs/aishell/s10/chain/egs_dataloader.py

```diff
@@ -0,0 +1,230 @@
+                    yield pseudo_epoch, batch
+
+
+class NnetChainExampleCollateFunc:
+
+    def __init__(self, egs_left_context, egs_right_context,
+                 frame_subsampling_factor=3):
+
+        '''
+        egs_left_context is from egs/info/left_context
+        egs_right_context is from egs/info/right_context
+        '''
+        assert egs_left_context >= 0
+        assert egs_left_context >= 0
+
+        # currently support either no subsampling or
+        # subsampling factor to be 3
+        assert frame_subsampling_factor in [1, 3]
+
+        self.egs_left_context = egs_left_context
+        self.egs_right_context = egs_right_context
+        self.frame_subsampling_factor = frame_subsampling_factor
+        
+    def __call__(self, eg):
+        '''
+        eg is a batch as it has been merged
+        '''
+        assert eg.inputs[0].name == 'input'
+        assert len(eg.outputs) == 1
+        assert eg.outputs[0].name == 'output'
+
+        # the length of below two return values will be 1,
+        # as we have merged egs into one batch
+
+        # TODO(haowen): remove the list wrapping
+        feature_list = []
+        supervision_list = []
+
+        supervision = eg.outputs[0].supervision
+        supervision_list.append(supervision)
+
+        batch_size = supervision.num_sequences
+        frames_per_sequence = (supervision.frames_per_sequence *
+                               self.frame_subsampling_factor) + \
+            self.egs_left_context + self.egs_right_context
+
+
+        _feats = kaldi.FloatMatrix()
+        eg.inputs[0].features.GetMatrix(_feats)
+        feats = _feats.numpy()
+
+        if len(eg.inputs) > 1:
+            _ivectors = kaldi.FloatMatrix()
+            eg.inputs[1].features.GetMatrix(_ivectors)
+            ivectors = _ivectors.numpy()
+
+        assert feats.shape[0] == batch_size * frames_per_sequence
+
+        feat_list = []
+        for i in range(batch_size):
+            start_index = i * frames_per_sequence
+            if self.frame_subsampling_factor == 3:
+                shift = np.random.choice([-1, 0, 1], 1)[0]
+                start_index += shift
+
+            end_index = start_index + frames_per_sequence
+            start_index += 2  # remove the leftmost frame added for frame shift
+            end_index -= 2  # remove the rightmost frame added for frame shift
+            feat = feats[start_index:end_index:, :]
+            if len(eg.inputs) > 1:
+                repeat_ivector = torch.from_numpy(
+                    ivectors[i]).repeat(feat.shape[0], 1)
+                feat = torch.cat(
+                    (torch.from_numpy(feat), repeat_ivector), dim=1).numpy()
+            feat_list.append(feat)
+
+        batched_feat = np.stack(feat_list, axis=0)
+        assert batched_feat.shape[0] == batch_size
+
+        # -4 = -2 -2
+        # the first -2 is from extra left/right context
+        # the second -2 is from lda feats splicing
```

Comment:
this comment is not correct any more.

Initially, line 172 is `start_index += 1`, now it is `+= 2`.

We can simply remove this comment.